pyext/src/output.py

```diff
@@ -691,57 +691,119 @@ def write_stats2(self):
     def __init__(self, filename):
         self.filename = filename
-        self.isstat1 = False
-        self.isstat2 = False
+        self.isstat1=False
+        self.isstat2=False
 
         # open the file
         if not self.filename is None:
             f = open(self.filename, "r")
         else:
             raise ValueError("No file name provided. Use -h for help")
 
-        # get the keys from the first line
-        for line in f.readlines():
-            d = eval(line)
-            self.klist = list(d.keys())
-            # check if it is a stat2 file
-            if "STAT2HEADER" in self.klist:
-                self.isstat2 = True
-                for k in self.klist:
-                    if "STAT2HEADER" in str(k):
-                        # if print_header: print k, d[k]
-                        del d[k]
-                stat2_dict = d
-                # get the list of keys sorted by value
-                kkeys = [k[0]
-                         for k in sorted(stat2_dict.items(), key=operator.itemgetter(1))]
-                self.klist = [k[1]
-                              for k in sorted(stat2_dict.items(), key=operator.itemgetter(1))]
-                self.invstat2_dict = {}
-                for k in kkeys:
-                    self.invstat2_dict.update({stat2_dict[k]: k})
-            else:
-                self.isstat1 = True
-                self.klist.sort()
+        self.exp_dict={} # Store all experimental header details here.
+
+        # get the {CategoryID:Category} pairs from the first line
+        line = f.readline()
+
+        #Store these keys in a dictionary. Example pair: {109 :'Total_Score'}
+        self.dict = self.parse_line(line)
+
+        self.klist = list(self.dict.keys())
+
+        if "STAT2HEADER" in self.klist:
+            self.isstat2 = True
+            for k in self.klist:
+                if "STAT2HEADER" in str(k):
+                    # if print_header: print k, d[k]
+                    self.exp_dict[k]=self.dict[k]
+                    del self.dict[k]
+
+            # get the list of keys sorted by value
+            kkeys = [k[0]
+                    for k in sorted(self.dict.items(), key=operator.itemgetter(1))]
+            self.klist = [k[1]
+                    for k in sorted(self.dict.items(), key=operator.itemgetter(1))]
+            self.inv_dict = {}
+
+            for k in kkeys:
+                self.inv_dict.update({self.dict[k]: k})
+        else:
+            print("WARNING: statfile v1 is deprecated.  Please convert to statfile v2")
+            self.isstat1 = True
+            self.klist.sort()
+        
 
-            break
         f.close()
 
+
+    def parse_line(self, line):
+        # Parses a line and returns a dictionary of key:value pairs
+        # {key:value, key:value, key:"{v1, v2, v3}", key:value}
+        d={} # output dictionary 
+        # First, remove outer braces and split via double quotes to isolate multi-component values.
```

Comment:
Splitting on double quotes won't work if multi-component values themselves contain any quotes (as the `STAT2HEADER_ENVIRON` value often does) since then the values may be single-quoted rather than double-quoted (and will contain `\'` and/or `\"` which you'll need to handle).